_posts/2020-07-04-high-performance-scheduling.md

```diff
@@ -0,0 +1,344 @@
+
+    Proposal: Python is slow.  Want to make it faster?  Don't use Python.  See
+    academic projects.
+
+    Challenge: This makes sense for some parts of the pipeline above, but not
+    for others.  It also makes it harder to attract maintainers.
+
+    What we should consider: Some parts of the scheduler and optimization
+    algorithms could be written in a lower level language, maybe Cython.  We'll
+    need to be careful about maintainability.
+
+2.  Distributed scheduling
+
+    Proposal: The scheduler is slow, maybe have many schedulers?  See Ray.
+
+    Challenge: It's actually really hard to make the kinds of decisions that
+    Dask has to make if scheduling state is spread on many computers.
+    Distributed scheduling works better when the workload is very either
+    uniform or highly decoupled.
+    Distributed scheduling is really attractive to people who like solving
+    interesting/hard problems.
+
+    What we should consider: We can move some simple logic down to the workers.
+    We've already done this with the easy stuff though.
+    It's not clear how much additional benefit there is here.
+
+3.  Build specialty scheduling around collections
+
+    Proposal: If Dask were to become just a dataframe library or just an array
+    computing library then it could special-case things more effectively.  See
+    Spark, Mars, and others.
+
+    Challenge: Yes, but Dask is not a dataframe library or an array library.
+    The three use cases we mention above are all very different.
+
+    What we should consider: modules like dask array and dask dataframe should
+    develop high level query blocks, and we should endeavor to
+    communicate these subgraphs over the wire directly so that they are more
+    compact.
+
+
+What should we actually do?
+---------------------------
+
+Because our pipeline has many stages, each of which can be slow for different
+reasons, we'll have to do many things.  Additionally, this is a hard problem
+because changing one piece of the project at this level has repurcussions for
+many other pieces.  The rest of this post tries to lay out a consistent set of
+changes.  Let's start with a summary:
+
+1.  For Dask array/dataframe let's use high level graphs more aggressively so
+    that we can communicate only abstract representations between the client
+    and scheduler.
+2.  But this breaks low level graph optimizations, fuse, cull, and slice fusion
+    in particular.  We can make these unnecessary with two changes:
+    -   We can make high level graphs considerably smarter to handle cull and slice fusion
+    -   We can move a bit more of the scheduling down to the workers to remove
+        the advantages of low-level fusion
+3.  Then, once all of the graph manipulation happens on the scheduler, let's
+    try to accelerate it, hopefully in a language that the current dev
+    community can understand, like Cython
+4.  At the same time in parallel, let's take a look at our network stack
+
```

Comment:
5. Provide more diagnostics on the dashboard for stages 3-6, which might be less work :)

This doesn't help with the first couple stages though.

Reply:
I'm curious about your thoughts on this.  In general I agree that visibility over this process would help us improve it.  Historically this has been hard because some of these stages run on the client side.  Now that we're moving more of it over to the scheduler hopefully it's more possible.  

Would this be just an bar/pie chart of costs, or something else?

Reply:
Off the top of my head, `<thing>.compute()` sends small messages to the scheduler with status updates

- "compute on `<thing>` requested"
- "optimized `<thing>` to ... tasks. Here it comes"
- "Sent `<thing>` in ...s"

And the dashboard represents these somehow (perhaps as a coarse progress bar).

But, these are sufficiently different from the broader goals in this post that I think we don't need to go into detail, and perhaps we ignore the suggestion entirely, if we think it distracts from the message.

Reply:
@jacobtomlinson  tried this in a dask/distributed PR.  It's tricky because these calls don't currently await within them reliably.  For example `client.submit` doesn't need an await.  So it's hard to send something up to the scheduler before we start serializing and such.  

Hopefully though a lot of this pain goes away once the early stages are all cheap, and most of the administrative cost occurs in the scheduler.

Reply:
That was here if you want to see the conversation dask/distributed#3363